### ISSUE TITLE
[FEATURE REQUEST] MIDI MTS SysEx

### ISSUE THREAD

**Issue by OpenSourceAnarchist** [spessasus/SpessaSynth#29]
This project is exactly what I was looking for. In short, I'm going to implement the dynamic tuning algorithm proposed in [this paper](https://doi.org/10.48550/arXiv.1706.04338) and implemented [here](https://gitlab.com/tp3/JustIntonation/-/blob/master/application/modules/tuner/tuneralgorithm.cpp). This requires having access to a list of currently playing notes as well as all previously played notes within the previous _n_ seconds, although this is easy enough for me to manually compute with your library. However, I need to set tunings for individual notes, and this is most easily accomplished with SysEx calls in the MIDI Tuning Standard.

See [this enum declaration](https://github.com/FluidSynth/fluidsynth/blob/b8fb6c81e1ca27c0bba2f6a0168832214f91d497/src/midi/fluid_midi.h#L195-L207) and [this API documentation](https://www.fluidsynth.org/api/group__tuning.html#ga3a9f995c967a9a227f9603ff57ee2304) for how Fluidsynth handles this. I can write my own functions to convert between Hz/cents/MIDI frequency data format, but without your library being able to respond to the SysEx calls, I don't know any other way to modify the tunings for all the MIDI notes, for just an octave and extrapolated to the rest, to individual notes, etc. More information about the MTS SysEx calls is available here: [http://www.microtonal-synthesis.com/MIDItuning.html](http://www.microtonal-synthesis.com/MIDItuning.html)

To be honest, I really like the way JZZ implements this as well as the helper functions. See [this code and the associated _helper functions](https://github.com/jazz-soft/JZZ/blob/master/javascript/JZZ.js#L1884-L1890) as well as the [API doc](https://jazz-soft.net/doc/JZZ/jzzmidi.html#sxNoteTuning).

I hope this was helpful enough to get an idea on why this is useful and how it could be implemented. Completely understand if this is on the backburner, and I'm open to any ideas you have as well :)

**Comment by spessasus** [comment]
Hi OpenSourceAnarchist,
I knew that midi tuning standard was a thing, but I never quite understood how it works.
I'm open to the idea, but I don't clearly understand it even with the helpul links you've provided.

Looking at the fluidsynth API, I don't really understand the "Tuning bank number" and "Tuning program number". Does it mean a tuning for every channel with a specific bank or a program? Or is it just a bank and program offset? And the notes: I understood that it's essentially an array of {note: tuning} in cents, but is this tuning relative or absolute?
Since you've mentioned that you can compute the tunings without a problem, I'm assuming that all of the "dump" messages can be omitted?
So, to sum it up,
you want to support the 
`MIDI_SYSEX_TUNING_NOTE_TUNE`
and
`MIDI_SYSEX_TUNING_NOTE_TUNE_BANK`?

I'll try to implement that, but without a test MIDI file it will be difficult to do so.
Anyways, thanks for the idea. :-)

PS: [Here's the current tuning calculation code](https://github.com/spessasus/SpessaSynth/blob/cf53b34dca3abf42eff42e3f85e3f4a40fdd6303/src/spessasynth_lib/synthetizer/worklet_system/worklet_methods/voice_control.js#L59) if that helps.

**Comment by spessasus** [comment]
I've added MTS octave tuning support. I will work on note tuning support now!

**Comment by OpenSourceAnarchist** [comment]
I JUST SAW YOUR COMMENT. Thank you so much for working on the MTS octave support. This might not be much help, but I was already writing it as you posted the reply :)

To be honest, my head has been spinning trying to read and remember the MIDI spec, the various additions, fluidsynth API, various js midi library APIs, web midi, and of course they all use similar terms that might be used totally differently in a different part of the spec!

From what I've gathered, MIDI tuning is in cents starting with C. So for standard 12 tone ET, starting with MIDI
note 0 (5 octaves below middle C), it would run 0, 100, 200, 300, ..., 127000 since MIDI (defaults?) to 12 TET so the MIDI note # times 100 gives you the cents. It is absolute.

I was under the impression the tuning bank and program numbers are completely independent of MIDI instrument banks or program numbers. I believe it's so performers can switch between a large number of premade tunings on-the-fly. [On the fluidsynth dev maling list](https://fluid-dev.nongnu.narkive.com/Y8FN2wzn/perfect-as-opposed-to-tempered-tuning) there is some more discussion on this, you can see they recommend keeping it easy and just modifying tuning bank 0 program 0.

The dump messages may be useful for other people, but for me specifically it wouldn't help. The two SysEx tuning commands you said would be helpful. Honestly you're better off looking at fluidsynth's implementation directly, see all the code for the enum cases [here](https://github.com/FluidSynth/fluidsynth/blob/b8fb6c81e1ca27c0bba2f6a0168832214f91d497/src/synth/fluid_synth.c#L2033). The `MIDI_SYSEX_TUNING_NOTE_TUNE` and `MIDI_SYSEX_TUNING_NOTE_TUNE_BANK` are implemented together.

The most useful things for me are 1) being able to store (or easily retrieve) the most recently played (unmodified!) notes, like "c4" or "midi note 60". I can manually implement this by keeping a running average of how many NoteOn events were seen for each MIDI note in the last X seconds/ticks, or you may have methods for querying active notes at a given time interval through the sequencer. And 2) being able to, in real-time, alter the cents/Hz of the actively voiced notes, including if they are still "On". I was thinking per-note MTS SysEx calls would be the easiest, but after looking at how you calculate tunings, it may be easier to interface directly with your code. I would basically add a function that would override the cents of the note before any of your existing tuning code that takes into account pitch bends on the channel, vibrato, etc. I'm not sure how easy this would be over carefully timed SysEx calls. Both should be straightforward as long as I can easily retrieve all the actively voiced MIDI note numbers.

**Comment by spessasus** [comment]
Hi,
I've added the note tuning support. (`MIDI_SYSEX_TUNING_NOTE_TUNE` and `MIDI_SYSEX_TUNING_NOTE_TUNE_BANK`)

Though, I only found one MIDI to test it with ([from this issue](https://github.com/FluidSynth/fluidsynth/issues/102)), so I can't guarantee that it will work every time.

I'm keeping this issue open. If the tuning works fine, please close it.
Good luck with your project!

**Comment by OpenSourceAnarchist** [comment]
I appreciate this a lot truly. I will work on hacking together an implementation of the tuning algorithm with your library and see if any issues or questions come up! I'm so glad I stumbled on your reddit post and repo. Batteries included with MIDI on web is so refreshing <3

**Comment by OpenSourceAnarchist** [comment]
Just a few questions/comments on the implementation:
I'm a little concerned that single note change SysEx calls aren't getting the right priority in the tuning calculation. In `src/spessasynth_lib/synthetizer/worklet_system/worklet_methods/voice_control.js`, shouldn't the MTS tuning be applied and take full precedence _after_ the more global parameters, e.g. MIDI master tuning, coarse grain, fine tune, scale tunings/channel tunings? I think the MTS spec is ambiguous enough that it's up to you, but I would expect any real-time MTS SysEx messages I send to override any semitone/cents modifications or custom controllers other than the vibrato.

What do you think about something like this instead? In general, can you think of any other interactions or reasons the final pitch might deviate from the MTS frequency?

```
    // TUNING
    let targetKey = voice.targetKey;

    // calculate tuning
    let cents = voice.modulatedGenerators[generatorTypes.fineTune]
        + channel.customControllers[customControllers.channelTuning]
        + channel.customControllers[customControllers.channelTransposeFine]
        + channel.customControllers[customControllers.masterTuning]
        + channel.channelOctaveTuning[voice.midiNote % 12];
    let semitones = voice.modulatedGenerators[generatorTypes.coarseTune]
        + channel.customControllers[customControllers.channelTuningSemitones];

    // calculate tuning by key
    cents += (targetKey - voice.sample.rootKey) * voice.modulatedGenerators[generatorTypes.scaleTuning];

    // midi tuning standard, overrides other global tuning parameters
    const tuning = this.tunings[channel.preset.program]?.[targetKey];
    if(tuning?.midiNote >= 0)
    {
        cents = tuning.centTuning;
        semitones = 0;
    }
    ...
```
In general, I just want to make sure no other parts of the library assume 12 tone equal temperament. I'm unsure how this actually plays out with rendering samples, e.g. if there's only one octave and it needs to be pitch shifted. MTS always seemed ad hoc.
Oh, and pitch bends are tricky too. Should they take precedence over an MTS tuning? To me, I would expect pitch bends to be overridden by MTS, but I can see an argument for the other case. For my case, I would want to put the logic of handling pitch bends in my own MTS code, e.g. applying a channel pitch bend amount to the cents of the original MIDI note, and *then* modifying it according to my custom tuning algorithm. If you want pitch bends to take precedence, I can just manually ignore them if my tuning algorithm is being applied.

**Comment by spessasus** [comment]
> Just a few questions/comments on the implementation: I'm a little concerned that single note change SysEx calls aren't getting the right priority in the tuning calculation. In `src/spessasynth_lib/synthetizer/worklet_system/worklet_methods/voice_control.js`, shouldn't the MTS tuning be applied and take full precedence _after_ the more global parameters, e.g. MIDI master tuning, coarse grain, fine tune, scale tunings/channel tunings? I think the MTS spec is ambiguous enough that it's up to you, but I would expect any real-time MTS SysEx messages I send to override any semitone/cents modifications or custom controllers other than the vibrato.
> 
> What do you think about something like this instead? In general, can you think of any other interactions or reasons the final pitch might deviate from the MTS frequency?
> 
> ```
>     // TUNING
>     let targetKey = voice.targetKey;
> 
>     // calculate tuning
>     let cents = voice.modulatedGenerators[generatorTypes.fineTune]
>         + channel.customControllers[customControllers.channelTuning]
>         + channel.customControllers[customControllers.channelTransposeFine]
>         + channel.customControllers[customControllers.masterTuning]
>         + channel.channelOctaveTuning[voice.midiNote % 12];
>     let semitones = voice.modulatedGenerators[generatorTypes.coarseTune]
>         + channel.customControllers[customControllers.channelTuningSemitones];
> 
>     // calculate tuning by key
>     cents += (targetKey - voice.sample.rootKey) * voice.modulatedGenerators[generatorTypes.scaleTuning];
> 
>     // midi tuning standard, overrides other global tuning parameters
>     const tuning = this.tunings[channel.preset.program]?.[targetKey];
>     if(tuning?.midiNote >= 0)
>     {
>         cents = tuning.centTuning;
>         semitones = 0;
>     }
>     ...
> ```
> 
> In general, I just want to make sure no other parts of the library assume 12 tone equal temperament. I'm unsure how this actually plays out with rendering samples, e.g. if there's only one octave and it needs to be pitch shifted. MTS always seemed ad hoc. Oh, and pitch bends are tricky too. Should they take precedence over an MTS tuning? To me, I would expect pitch bends to be overridden by MTS, but I can see an argument for the other case. For my case, I would want to put the logic of handling pitch bends in my own MTS code, e.g. applying a channel pitch bend amount to the cents of the original MIDI note, and _then_ modifying it according to my custom tuning algorithm. If you want pitch bends to take precedence, I can just manually ignore them if my tuning algorithm is being applied.
> 
> EDIT: In the MTS Spec, they say:
> 
> > Recommended Practice (RP-020)
> > Defaults for Scale/Octave Tuning
> > If tuning presets are not supported by the instrument, it is assumed that the initial tuning of the instrument is equal
> > temperament. If presets are supported, it is suggested that the first preset, selected by Bank 0H and Preset 0H, would be
> > equal temperament. Tuning adjusters should begin by selecting Bank 0H, Preset 0H in order to start from equal
> > temperament, if that is the desired behavior.
> 
> Perhaps it would be easier to just always read from 0,0 and predefine ET tunings (midi note # * 100), that way you avoid the conditional and can just look up the values for the pitch directly from the bank/program tuning.

I implemented mts the same way that TiMidity++ does (as fluidsynth doesn't seem to work because of that bug i got the mts.mid from)

Your code wouldn't work at all, because:
[take a look at this](https://github.com/spessasus/SpessaSynth/blob/3eb38882dcb25a0a5bda21cc479ee30e18124cfe/src/spessasynth_lib/synthetizer/worklet_system/worklet_methods/note_on.js#L41)
The `this.tunings` table has two properties: midiNote and centTuning.
- The midiNote is like "coarse tune". It shows the base MIDI note to use (the first byte of the MTS frequency). 
- Cent tuning is just the fine tune addition to that note being used, maxing out at 100 and min -100. The semitones being forced to 0 cause the midiNote from tuning to be ignored, causing the note to have the incorrect tuning.

Not to mention that discarding the tunings is catastrophic because:
1. We discard soundfont's tuning generators which is a big no-no. These ALWAYS have to apply. It's not controllable by MIDI and it ensures that the sample plays correctly.
2. Every sample has a specific "original key" number which is used to determine the proper pitch (this line: `(targetKey - voice.sample.rootKey) * voice.modulatedGenerators[generatorTypes.scaleTuning];`) Setting semitones to 0 will cause the sample to sound like it's directly stored in the file. This effectively kills ALL tuning (including MTS) completely. Currently, MTS overrides the "original key" which tunes the sample to the desired frequency. But setting semitones to 0 overrides that, and everything else.

- If you want to avoid pitch wheel, simply set pitch bend range to 0 via RPN.
- If you want to avoid coarseTune and fineTune RPNs, simply don't set them. They are always set to 0 by default.
- You can't avoid vibrato, as it's the part of the sf2 standard. You can set the modulation wheel to 0, but if the soundfont has vibrato hard-coded, it will be on. But most of them just use the mod wheel so that solution would work.
- Custom controllers are only applied with the synth's functions, and they are always set to 0 by default. Simply don't change them and everything should work as expected.

**Comment by OpenSourceAnarchist** [comment]
Yes, your implementation and precedence/ordering make more sense now. I didn't understand the purpose of `midiNote` in `this.tunings`. I thought `midiNote` was referring to the _original_ midi note number, and that `centTuning` was the total deviation from the standard ET pitch in cents. Of course it is obvious when looking at `const tuning = this.tunings[channel.preset.program]?.[targetKey];`, sorry for the confusion :)

I believe I understand the chain now, thank you again! I'll definitely learn with some edge cases and sanity checking the final pitches. But this seems much simpler and I can handle all the different situations in my own logic.

**Comment by spessasus** [comment]
Hi @OpenSourceAnarchist,
How is the tuning algorithm going? This issue has been open for around a month now, so if the MTS works fine, I'll close it. Let me know!

**Comment by OpenSourceAnarchist** [comment]
I'm not done with it, but if I find any issues I'll reopen this. Otherwise I'll just let you know when it's done so you can check it out!
I just moved so I've been adjusting to new routines and left the last part unfinished :(
Thanks for implementing this again, it means a lot!

**Comment by paulikauro** [comment]
Thanks for implementing this (and the whole library!) :)

I think there is a bug in the single note tuning :thinking: . `noteOn` translates the MIDI note using `this.tunings` to the "coarse tune" but the original note number (that was passed to `noteOn`) seems to get lost at this point, which means that the tuning calculation in `renderVoice` is using the already translated MIDI note (coarse tune) to look up the cent tuning from `this.tunings`, which will only work if the coarse tune matches the original MIDI note (and if there's no entry in `this.tunings` for the coarse tune).

For example, this works and produces a justly tuned major third ("MIDI pitches" 40 and 43.86) as expected:
```js
const remap = (key, pitch) => {
    let pitch_key = Math.floor(pitch);
    const frac = Math.floor((pitch - pitch_key) / 0.000061);
    synth.systemExclusive([0x7F, 0x7F, 0x08, 0x02, 0, 1, key, pitch_key, (frac >> 7) & 0x7F, frac & 0x7F]);
};

remap(1, 40);
remap(43, 43.86); // coarse tune matches the original MIDI note
synth.noteOn(0, 1, 127);
synth.noteOn(0, 43, 127);
```
However this doesn't and produces a 12TET minor third (40 and 43) instead:
```js
remap(1, 40);
remap(2, 43.86);
synth.noteOn(0, 1, 127);
synth.noteOn(0, 2, 127);
```
And this produces a weird-sounding middle C (60) instead of 40 (I guess because of the soundfont things you mentioned):
```js
remap(1, 40);
remap(40, 60);
synth.noteOn(0, 1, 127);
```

(Btw, `index.js` seems to have two broken imports in v3.21.0: `modulatorSources` and `NON_CC_INDEX_OFFSET`)

**Comment by spessasus** [comment]
Hi @paulikauro ,

I've fixed the imports in the npm package, thanks for that!

I'll look into the MTS later.

**Comment by spessasus** [comment]
Hi,
I've fixed it (I think). The tunings would override each other, so I hope it works now.

**Comment by paulikauro** [comment]
Yep, everything I tried and need works now. Thanks a lot for the quick fixes!
